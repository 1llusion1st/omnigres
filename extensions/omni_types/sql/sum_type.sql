\pset null 'NULL'
\set dump_types 'select typname,variants from omni_types.sum_types st inner join pg_type t on t.oid = st.oid'

create or replace function get_type_size(data_type regtype)
    returns integer as
$$
select
    typlen
from
    pg_type
where
    oid = data_type
$$
    language sql;


-- Empty
begin;

select omni_types.sum_type('empty', variadic array []::regtype[]);
\dT
:dump_types;

select null::empty is null as is_null;

rollback;

-- Single fixed size by val
begin;
select omni_types.sum_type('sum_type', 'integer');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'integer(100)'::sum_type;
select 'integer(100)'::sum_type::integer;
select 100::sum_type;

rollback;

-- Multiple fixed size by val
begin;
select omni_types.sum_type('sum_type', 'integer', 'bigint');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'integer(1000)'::sum_type;
select 'integer(1000)'::sum_type::integer;
select 1000::sum_type;
select 'bigint(10000000000)'::sum_type;
select 'bigint(10000000000)'::sum_type::bigint;
select 10000000000::bigint::sum_type;

rollback;

-- Single fixed size
begin;
select omni_types.sum_type('sum_type', 'name');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'name(test)'::sum_type;
select 'name(test)'::sum_type::name;
select 'test'::name::sum_type;

rollback;

-- Multiple fixed size, by val is mixed
begin;
select omni_types.sum_type('sum_type', 'name', 'integer');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'name(test)'::sum_type;
select 'name(test)'::sum_type::name;
select 'test'::name::sum_type;
select 'integer(1000)'::sum_type;
select 'integer(1000)'::sum_type::integer;
select 1000::sum_type;

rollback;

-- Single variable size
begin;
select omni_types.sum_type('sum_type', 'text');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'text(Hello)'::sum_type;
select 'text(Hello)'::sum_type::text;
select 'Hello'::text::sum_type;

rollback;

-- Multiple mixed variable and fixed size
begin;
select omni_types.sum_type('sum_type', 'text', 'integer', 'name');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'text(Hello)'::sum_type;
select 'text(Hello)'::sum_type::text;
select 'Hello'::text::sum_type;
select 'integer(1000)'::sum_type;
select 'integer(1000)'::sum_type::integer;
select 1000::sum_type;
select 'name(test)'::sum_type;
select 'name(test)'::sum_type::name;
select 'test'::name::sum_type;

rollback;

-- Domains
begin;
create domain height as integer check (value > 0);
create domain age as integer check (value > 0 and value < 200);

select omni_types.sum_type('sum_type', 'height', 'age');
\dT;
:dump_types;

select get_type_size('sum_type');

select 'height(100)'::sum_type;
select 'age(100)'::sum_type;

-- Can't cast domain types, use functions
select sum_type_from_height(100);
select sum_type_from_age(100);

rollback;

--- Composite types
begin;
create type person as
(
    name text,
    dob  date
);

create type animal as enum ('dog', 'cat', 'fish', 'other');

create type pet as
(
    name   text,
    animal animal
);

select omni_types.sum_type('sum_type', 'person', 'pet');
\dT;
:dump_types;

select get_type_size('sum_type');

select $$person((John,01/01/1980))$$::sum_type;
select $$pet((Charlie,dog))$$::sum_type;
select row ('John', '01/01/1980')::person::sum_type;
select row ('Charlie','dog')::pet::sum_type;
select row ('John', '01/01/1980')::person::sum_type::person;
select row ('John', '01/01/1980')::person::sum_type::pet;
select row ('Charlie','dog')::pet::sum_type::pet;
select row ('Charlie','dog')::pet::sum_type::person;


rollback;

-- Casting variants to different types
begin;
select omni_types.sum_type('sum_type', 'integer', 'boolean');
\dT;
:dump_types;

select 100::sum_type::boolean;
select true::sum_type::integer;

rollback;

-- Duplicates
begin;
select omni_types.sum_type('sum_type', 'integer', 'integer');

rollback;

-- Ensure no types are leaked
\dT;
:dump_types;