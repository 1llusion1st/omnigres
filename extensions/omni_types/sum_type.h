// clang-format off
#include <postgres.h>
#include <fmgr.h>
// clang-format on

typedef int32 Discriminant;
// This is to basically ensure a theoretical limit of types can be covered by the discriminant
// (although it is generally excessive)
StaticAssertDecl(sizeof(uint32_t) <= sizeof(Oid),
                 "discriminant size should not be able to cover more than there can be types");
StaticAssertDecl(sizeof(uint32_t) == sizeof(Oid), "discriminant size should match that of Oid");

/**
 * Fixed-size type layout
 */
typedef struct {
  Discriminant discriminant;
  char data[FLEXIBLE_ARRAY_MEMBER];
} FixedSizeVariant;

/**
 * Variable-size type layout
 */
typedef struct {
  Discriminant discriminant;
  /**
   * @note If the variant iself is a `varlena`, `data` will embed that `varlena` in its entirety
   * to make it easier to work with (no need to create a new varlena with the reduced size).
   * It increases the size of the overall structure by the size of  the `varlena` header.
   */
  struct varlena data;
} VarSizeVariant;

StaticAssertDecl(offsetof(VarSizeVariant, discriminant) == offsetof(FixedSizeVariant, discriminant),
                 "discriminant offset should be equal");
