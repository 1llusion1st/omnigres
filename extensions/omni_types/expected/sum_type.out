\pset null 'NULL'
\set dump_types 'select typ as typname,variants from omni_types.sum_types'
create or replace function get_type_size(data_type regtype)
    returns integer as
$$
select
    typlen
from
    pg_type
where
    oid = data_type
$$
    language sql;
-- Empty
begin;
select omni_types.sum_type('empty', variadic array []::regtype[]);
 sum_type 
----------
 empty
(1 row)

\dT
      List of data types
 Schema | Name  | Description 
--------+-------+-------------
 public | empty | 
(1 row)

:dump_types;
 typname | variants 
---------+----------
 empty   | {}
(1 row)

select null::empty is null as is_null;
 is_null 
---------
 t
(1 row)

rollback;
-- Single fixed size by val
begin;
select omni_types.sum_type('sum_type', 'integer');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  | variants  
----------+-----------
 sum_type | {integer}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
             8
(1 row)

select 'integer(100)'::sum_type;
   sum_type   
--------------
 integer(100)
(1 row)

select 'integer(100)'::sum_type::integer;
 int4 
------
  100
(1 row)

select 100::sum_type;
   sum_type   
--------------
 integer(100)
(1 row)

rollback;
-- Multiple fixed size by val
begin;
select omni_types.sum_type('sum_type', 'integer', 'bigint');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |     variants     
----------+------------------
 sum_type | {integer,bigint}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            12
(1 row)

select 'integer(1000)'::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

select 'integer(1000)'::sum_type::integer;
 int4 
------
 1000
(1 row)

select 1000::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

select 'bigint(10000000000)'::sum_type;
      sum_type       
---------------------
 bigint(10000000000)
(1 row)

select 'bigint(10000000000)'::sum_type::bigint;
    int8     
-------------
 10000000000
(1 row)

select 10000000000::bigint::sum_type;
      sum_type       
---------------------
 bigint(10000000000)
(1 row)

rollback;
-- Single fixed size
begin;
select omni_types.sum_type('sum_type', 'name');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  | variants 
----------+----------
 sum_type | {name}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            68
(1 row)

select 'name(test)'::sum_type;
  sum_type  
------------
 name(test)
(1 row)

select 'name(test)'::sum_type::name;
 name 
------
 test
(1 row)

select 'test'::name::sum_type;
  sum_type  
------------
 name(test)
(1 row)

rollback;
-- Multiple fixed size, by val is mixed
begin;
select omni_types.sum_type('sum_type', 'name', 'integer');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |    variants    
----------+----------------
 sum_type | {name,integer}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            68
(1 row)

select 'name(test)'::sum_type;
  sum_type  
------------
 name(test)
(1 row)

select 'name(test)'::sum_type::name;
 name 
------
 test
(1 row)

select 'test'::name::sum_type;
  sum_type  
------------
 name(test)
(1 row)

select 'integer(1000)'::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

select 'integer(1000)'::sum_type::integer;
 int4 
------
 1000
(1 row)

select 1000::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

rollback;
-- Single variable size
begin;
select omni_types.sum_type('sum_type', 'text');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  | variants 
----------+----------
 sum_type | {text}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            -1
(1 row)

select 'text(Hello)'::sum_type;
  sum_type   
-------------
 text(Hello)
(1 row)

select 'text(Hello)'::sum_type::text;
 text  
-------
 Hello
(1 row)

select 'Hello'::text::sum_type;
  sum_type   
-------------
 text(Hello)
(1 row)

rollback;
-- Multiple mixed variable and fixed size
begin;
select omni_types.sum_type('sum_type', 'text', 'integer', 'name');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |      variants       
----------+---------------------
 sum_type | {text,integer,name}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            -1
(1 row)

select 'text(Hello)'::sum_type;
  sum_type   
-------------
 text(Hello)
(1 row)

select 'text(Hello)'::sum_type::text;
 text  
-------
 Hello
(1 row)

select 'Hello'::text::sum_type;
  sum_type   
-------------
 text(Hello)
(1 row)

select 'integer(1000)'::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

select 'integer(1000)'::sum_type::integer;
 int4 
------
 1000
(1 row)

select 1000::sum_type;
   sum_type    
---------------
 integer(1000)
(1 row)

select 'name(test)'::sum_type;
  sum_type  
------------
 name(test)
(1 row)

select 'name(test)'::sum_type::name;
 name 
------
 test
(1 row)

select 'test'::name::sum_type;
  sum_type  
------------
 name(test)
(1 row)

rollback;
-- Domains
begin;
create domain height as integer check (value > 0);
create domain age as integer check (value > 0 and value < 200);
select omni_types.sum_type('sum_type', 'height', 'age');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | age      | 
 public | height   | 
 public | sum_type | 
(3 rows)

:dump_types;
 typname  |   variants   
----------+--------------
 sum_type | {height,age}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
             8
(1 row)

select 'height(100)'::sum_type;
  sum_type   
-------------
 height(100)
(1 row)

select 'age(100)'::sum_type;
 sum_type 
----------
 age(100)
(1 row)

-- Can't cast domain types, use functions
select sum_type_from_height(100);
 sum_type_from_height 
----------------------
 height(100)
(1 row)

select sum_type_from_age(100);
 sum_type_from_age 
-------------------
 age(100)
(1 row)

rollback;
--- Composite types
begin;
create type person as
(
    name text,
    dob  date
);
create type animal as enum ('dog', 'cat', 'fish', 'other');
create type pet as
(
    name   text,
    animal animal
);
select omni_types.sum_type('sum_type', 'person', 'pet');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | animal   | 
 public | person   | 
 public | pet      | 
 public | sum_type | 
(4 rows)

:dump_types;
 typname  |   variants   
----------+--------------
 sum_type | {person,pet}
(1 row)

select get_type_size('sum_type');
 get_type_size 
---------------
            -1
(1 row)

select $$person((John,01/01/1980))$$::sum_type;
         sum_type          
---------------------------
 person((John,01-01-1980))
(1 row)

select $$pet((Charlie,dog))$$::sum_type;
      sum_type      
--------------------
 pet((Charlie,dog))
(1 row)

select row ('John', '01/01/1980')::person::sum_type;
            row            
---------------------------
 person((John,01-01-1980))
(1 row)

select row ('Charlie','dog')::pet::sum_type;
        row         
--------------------
 pet((Charlie,dog))
(1 row)

select row ('John', '01/01/1980')::person::sum_type::person;
        row        
-------------------
 (John,01-01-1980)
(1 row)

select row ('John', '01/01/1980')::person::sum_type::pet;
 row  
------
 NULL
(1 row)

select row ('Charlie','dog')::pet::sum_type::pet;
      row      
---------------
 (Charlie,dog)
(1 row)

select row ('Charlie','dog')::pet::sum_type::person;
 row  
------
 NULL
(1 row)

rollback;
-- Casting variants to different types
begin;
select omni_types.sum_type('sum_type', 'integer', 'boolean');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |     variants      
----------+-------------------
 sum_type | {integer,boolean}
(1 row)

select 100::sum_type::boolean;
 bool 
------
 NULL
(1 row)

select true::sum_type::integer;
 int4 
------
 NULL
(1 row)

rollback;
-- Duplicates
begin;
select omni_types.sum_type('sum_type', 'integer', 'integer');
ERROR:  Sum types can not contain duplicate variants
CONTEXT:  PL/pgSQL function omni_types.sum_type_unique_variants_trigger_func() line 12 at RAISE
SQL statement "insert into omni_types.sum_types (typ, variants) values ($1, $2)"
rollback;
-- Determining variant
begin;
select omni_types.sum_type('sum_type', 'integer', 'boolean');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |     variants      
----------+-------------------
 sum_type | {integer,boolean}
(1 row)

select omni_types.variant(100::sum_type);
 variant 
---------
 integer
(1 row)

select omni_types.variant(true::sum_type);
 variant 
---------
 boolean
(1 row)

-- Invalid type
select omni_types.variant(10);
 variant 
---------
 NULL
(1 row)

rollback;
-- TOASTing
begin;
select omni_types.sum_type('sum_type', 'text', 'integer');
 sum_type 
----------
 sum_type
(1 row)

\dT;
       List of data types
 Schema |   Name   | Description 
--------+----------+-------------
 public | sum_type | 
(1 row)

:dump_types;
 typname  |    variants    
----------+----------------
 sum_type | {text,integer}
(1 row)

create table test (val sum_type);
alter table test alter val set storage external;
insert into test values (repeat('a', 100000)::text::sum_type),(1000);
select omni_types.variant(val) from test;
 variant 
---------
 text
 integer
(2 rows)

select (case when omni_types.variant(val) = 'text'::regtype then length(val::text)
             when omni_types.variant(val) = 'integer'::regtype then val::integer
             else null end) from test;
  case  
--------
 100000
   1000
(2 rows)

rollback;
-- Pseudo-types can't be used
begin;
select omni_types.sum_type('sum_type', 'text', 'anyarray');
ERROR:  Pseudo-types can't be variants
DETAIL:  anyarray
rollback;
-- Ensure no types are leaked
\dT;
     List of data types
 Schema | Name | Description 
--------+------+-------------
(0 rows)

:dump_types;
 typname | variants 
---------+----------
(0 rows)

